rspirv/dr/build/autogen_type.rs

```diff
@@ -3,86 +3,130 @@
 // DO NOT MODIFY!
 
 impl Builder {
-    #[doc = "Appends an OpTypeVoid instruction and returns the result id."]
+    #[doc = "Appends an OpTypeVoid instruction and returns the result id, or return the existing id if the instruction was already present."]
     pub fn type_void(&mut self) -> spirv::Word {
+        let mut inst = dr::Instruction::new(spirv::Op::TypeVoid, None, None, vec![]);
+        for ty in &self.module.types_global_values {
+            if ty.is_type_identical(&inst) {
+                if let Some(id) = ty.result_id {
+                    return id;
+                }
+            }
+        }
         let id = self.id();
-        self.module.types_global_values.push(dr::Instruction::new(
-            spirv::Op::TypeVoid,
-            None,
-            Some(id),
-            vec![],
-        ));
+        inst.result_id = Some(id);
+        self.module.types_global_values.push(inst);
         id
     }
-    #[doc = "Appends an OpTypeBool instruction and returns the result id."]
+    #[doc = "Appends an OpTypeBool instruction and returns the result id, or return the existing id if the instruction was already present."]
     pub fn type_bool(&mut self) -> spirv::Word {
+        let mut inst = dr::Instruction::new(spirv::Op::TypeBool, None, None, vec![]);
+        for ty in &self.module.types_global_values {
+            if ty.is_type_identical(&inst) {
+                if let Some(id) = ty.result_id {
+                    return id;
+                }
+            }
+        }
         let id = self.id();
-        self.module.types_global_values.push(dr::Instruction::new(
-            spirv::Op::TypeBool,
-            None,
-            Some(id),
-            vec![],
-        ));
+        inst.result_id = Some(id);
+        self.module.types_global_values.push(inst);
         id
     }
-    #[doc = "Appends an OpTypeInt instruction and returns the result id."]
+    #[doc = "Appends an OpTypeInt instruction and returns the result id, or return the existing id if the instruction was already present."]
     pub fn type_int(&mut self, width: u32, signedness: u32) -> spirv::Word {
-        let id = self.id();
-        self.module.types_global_values.push(dr::Instruction::new(
+        let mut inst = dr::Instruction::new(
             spirv::Op::TypeInt,
             None,
-            Some(id),
+            None,
             vec![
                 dr::Operand::LiteralInt32(width),
                 dr::Operand::LiteralInt32(signedness),
             ],
-        ));
+        );
+        for ty in &self.module.types_global_values {
```

Comment:
if we are to add de-duplication (which sounds iffy to me), we'd at least want that code shared, i.e. have a command function to add a type instruction that we'd call from these, and which would try to deduplicate

Reply:
I can make that happen :-)